tests/subsys/pm/device_runtime/src/main.c

```diff
@@ -0,0 +1,237 @@
+/*
+ * Copyright (c) 2021 Intel Corporation.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr.h>
+#include <ztest.h>
+#include <kernel.h>
+#include <pm/pm.h>
+#include <pm/device_runtime.h>
+#include "dummy_driver.h"
+
+#define MAX_TIMES 10
+#define STACKSIZE 1024
+
+/* Semaphore used to synchronize thread A and thread B*/
+K_SEM_DEFINE(sem, 0, 1);
+
+K_THREAD_STACK_DEFINE(threadA_stack, STACKSIZE);
+K_THREAD_STACK_DEFINE(threadB_stack, STACKSIZE);
+
+static const struct device *dev;
+static struct dummy_driver_api *api;
+
+static struct k_thread threadA;
+static struct k_thread threadB;
+
+
+void threadA_func(void *arg1, void *arg2, void *arg3)
+{
+	int ret;
+
+	ARG_UNUSED(arg1);
+	ARG_UNUSED(arg2);
+	ARG_UNUSED(arg3);
+
+	ret = api->open(dev);
+	zassert_true(ret == 0, "Fail to get device");
+
+	/* Lets allow threadB run */
+	k_sem_give(&sem);
+
+	/* Block waiting for device operation conclude */
+	ret = api->wait(dev);
+	zassert_true(ret == 0, "Fail to wait transaction");
+
+	/* At this point threadB should have put the device and
+	 * the current state should be SUSPENDED.
+	 */
+	zassert_true(dev->pm->state == PM_DEVICE_STATE_SUSPEND, "Wrong state");
+
+	k_sem_take(&sem, K_FOREVER);
+
+	ret = api->open(dev);
+	zassert_true(ret == 0, "Fail to get device");
+
+	/* Lets allow threadB run */
+	k_sem_give(&sem);
+
+	ret = api->wait(dev);
+	zassert_true(ret == 0, "Fail to wait transaction");
+
+	zassert_true(dev->pm->state == PM_DEVICE_STATE_ACTIVE, "Wrong state");
+}
+
+void threadB_func(void *arg1, void *arg2, void *arg3)
+{
+	int ret;
+
+	ARG_UNUSED(arg1);
+	ARG_UNUSED(arg2);
+	ARG_UNUSED(arg3);
+
+	k_sem_take(&sem, K_FOREVER);
+
+	api->close(dev);
+
+	k_sem_give(&sem);
+	ret = api->wait(dev);
+	zassert_true(ret == 0, "Fail to wait transaction");
+
+	/* Check the state */
+	zassert_true(dev->pm->state == PM_DEVICE_STATE_SUSPEND, "Wrong state");
+}
+
+/*
+ * @brief test device runtime concurrency
+ *
+ * @details
+ *  - Two threads will do different operations on a device. ThreadA will
+ *    try to bring up the device using an async call and then will be scheduled
+ *    out and threadB will run. ThreadB then will suspend the device and give up
+ *    in favor of threadA. At this point the device should reflect these
+ *    operations and be suspended.
+ *
+ * @see pm_device_get(), pm_device_put()
+ *
+ * @ingroup power_tests
+ */
+void test_concurrency(void)
+{
+	k_thread_start(&threadA);
+	k_thread_start(&threadB);
+
+	k_thread_join(&threadA, K_FOREVER);
+	k_thread_join(&threadB, K_FOREVER);
+}
+
+void test_setup(void)
+{
+	dev = device_get_binding(DUMMY_DRIVER_NAME);
+	api = (struct dummy_driver_api *)dev->api;
+
+	k_thread_create(&threadA, threadA_stack,
+			K_THREAD_STACK_SIZEOF(threadA_stack),
+			threadA_func, NULL, NULL, NULL,
+			K_PRIO_PREEMPT(1), 0, K_FOREVER);
+
+	/* Lets make threadB has higher priority than the workqueue
```

Comment:
why do we need to worry about this? This seems like a potential for problems if we have to worry about system workq priority, we should test threads at different, but random priorities and not be concerned about workq

Reply:
Otherwise we don't test concurrence, the workqueue will always run before the others threads .. Just because of this, if I don't do this we are not testing this corner case.